collections/_article/migrate-animation-from-godot-4-0-to-4-3.md

```diff
@@ -0,0 +1,317 @@
+
+If it is nested, the parent AnimationNode will consider the Nested StateMachine to be finished if it is in the End state, or if the State's transition destination does not exist and it is a dead end. If the parent AnimationNode commands the Nested StateMachine to play from the beginning, the animation in the current State is sequenced to position 0.
+
+This is mainly used in cases where states are not connected by a transition. That is, it can be used to reproduce the behavior of an Any state.
+
+![nested statemachine any state](/storage/blog/migrate-animation-from-godot-4-0-to-4-3/nested_state_machine-01.webp)
+
+#### Grouped
+
+This can be used to visually organize the States to avoid bloat in the StateMachine.
+
+Grouped StateMachines are not allowed to be played directly by the user using the playback object, but must be played from the parent StateMachine. This means that it must always have a Root or Nested StateMachine as an ancestor.
+
+Here is a diagram of the structure and behavior:
+
+![grouped statemachine](/storage/blog/migrate-animation-from-godot-4-0-to-4-3/grouped_state_machine-01.webp)
+
+**For example, if you want to play Attack2 in the above case, the script could be written as follows:**
+
+```gdscript
+var state_machine: AnimationNodeStateMachinePlayback = $AnimationTree.get("parameters/playback")
+state_machine.play("Attack/Attack2")
+```
+
+Also, `Grouped` type never stays in the `Start` state and `End` state. The `Start` state of a `Grouped` type is equal to the next state after the `Start` state of the state machine as seen by the parent state machine. The `End` state of a `Grouped` type is equal to the next state of the state machine in the parent state machine. In other words, those transitions are replaced by the parent state machine transitions.
+
+**In short, a more direct representation of the above diagram is equivalent to the following:**
+
+![omitted grouped statemachine](/storage/blog/migrate-animation-from-godot-4-0-to-4-3/grouped_state_machine_2-01.webp)
+
+Note that currently StateMachine does not have user-controllable indexes for connections and cannot have multiple Start/End states, so Grouped StateMachine is limited to having only one input/output. This limitation may be removed in the future by StateMachine enhancements.
+
+### NodeOneShot / NodeTransiton / StateMachineTransition - Break Loop at End
+
+![break loop property](/storage/blog/migrate-animation-from-godot-4-0-to-4-3/prop_break_loop.webp)
+
+In the 4.1 StateMachine rework, we make the AnimationNode has an infinite-length remain time in a loop and for StateMachines with unpredictable endings as a hack, since old AnimationNode only returned the remaining time after the processing but there was no information there about whether it was looping or not. This change was necessary to prevent unintended animation interruptions for consistency, but it removed the existing behavior of breaking the loop at `AtEnd` and doing a Transition.
```

Comment:
```suggestion
In the 4.1 StateMachine rework, we made the AnimationNode have an infinite-length remain time in a loop and for StateMachines with unpredictable endings as a hack. We did this becauses the old AnimationNode only returned the remaining time after the processing but there was no information there about whether it was looping or not. This change was necessary to prevent unintended animation interruptions for consistency, but it removed the existing behavior of breaking the loop at `AtEnd` and doing a Transition.
```